doc/TAM/Granular-Counter-Subscription.md

```diff
@@ -0,0 +1,208 @@
+   switch_id,
+   attr_count,
+   sai_attr_list);
+
+// Example: Create telemetry type object to collect object stats
+sai_attr_list[0].id = SAI_TAM_TEL_TYPE_ATTR_TAM_TELEMETRY_TYPE;
+sai_attr_list[0].value.u32 = SAI_TAM_TELEMETRY_TYPE_OBJECT_STAT;
+
+sai_attr_list[1].id = SAI_TAM_TEL_TYPE_ATTR_REPORT_ID;
+sai_attr_list[1].value.oid = sai_tam_report_obj;
+
+attr_count = 2;
+sai_create_tam_tel_type_fn(
+    &sai_tam_tel_type_obj,
+    switch_id,
+    attr_count,
+    sai_attr_list);
+
+// Example: Create counter subscription(s) to collect queue length and watermark
+sai_attr_list[0].id = SAI_TAM_COUNTER_SUBSCRIPTION_ATTR_TEL_TYPE;
+sai_attr_list[0].value.u32 = sai_tam_tel_type_obj;
+
+sai_attr_list[1].id = SAI_TAM_COUNTER_SUBSCRIPTION_ATTR_OBJECT_ID;
+sai_attr_list[1].value.oid = sai_queue_obj;
+
+sai_attr_list[2].id = SAI_TAM_COUNTER_SUBSCRIPTION_ATTR_STAT_ID;
+sai_attr_list[2].value.u32 = SAI_QUEUE_STAT_CURR_OCCUPANCY_BYTES;
+
+sai_attr_list[3].id = SAI_TAM_COUNTER_SUBSCRIPTION_ATTR_LABEL;
+sai_attr_list[3].value.u64 = 1;
+
+attr_count = 4;
+sai_create_tam_counter_subscription_fn(
+    &sai_tam_subscription_obj1,
+    switch_id,
+    attr_count,
+    sai_attr_list);
+
+sai_attr_list[0].id = SAI_TAM_COUNTER_SUBSCRIPTION_ATTR_TEL_TYPE;
+sai_attr_list[0].value.u32 = sai_tam_tel_type_obj;
+
+sai_attr_list[1].id = SAI_TAM_COUNTER_SUBSCRIPTION_ATTR_OBJECT_ID;
+sai_attr_list[1].value.oid = sai_queue_obj;
+
+sai_attr_list[2].id = SAI_TAM_COUNTER_SUBSCRIPTION_ATTR_STAT_ID;
+sai_attr_list[2].value.u32 = SAI_QUEUE_STAT_WATERMARK_BYTES;
+
+sai_attr_list[3].id = SAI_TAM_COUNTER_SUBSCRIPTION_ATTR_LABEL;
+sai_attr_list[3].value.u64 = 2;
+
+attr_count = 4;
+sai_create_tam_counter_subscription_fn(
+    &sai_tam_subscription_obj2,
+    switch_id,
+    attr_count,
+    sai_attr_list);
+
+// Example: Create Telemetry object
+sai_attr_list[0].id = SAI_TAM_TELEMETRY_ATTR_TAM_TYPE_LIST;
+sai_attr_list[0].value.objlist.count = 1
+sai_attr_list[0].value.objlist.list[0] = sai_tam_tel_type_obj;
+
+sai_attr_list[1].id = SAI_TAM_TELEMETRY_ATTR_COLLECTOR_LIST
+sai_attr_list[1].value.objlist.count = 1;
+sai_attr_list[1].value.objlist.list[0] = collector_obj;
+
+attr_count = 2;
+sai_create_tam_telemetry_fn(
+    &sai_tam_telemetry_obj,
+    switch_id,
+    attr_count,
+    sai_attr_list);
+
+// Example: Create TAM object and bind to monitored objects:
+
+sai_attr_list[0].id = SAI_TAM_ATTR_TAM_TELEMETRY_OBJECTS_LIST;
+sai_attr_list[0].value.objlist.count = 1;
+sai_attr_list[0].value.objlist.list[0] = sai_tam_telemetry_obj;
+
+sai_attr_list[1].id = SAI_TAM_ATTR_TAM_BIND_POINT_TYPE_LIST;
+sai_attr_list[1].value.objlist.count = 1;
+sai_attr_list[1].value.objlist.list[0] = SAI_TAM_BIND_POINT_TYPE_QUEUE;
+
+attr_count = 2;
+sai_create_tam_fn(
+    &sai_tam_obj,
+    switch_id,
+    attr_count,
+    sai_attr_list);
+
+// Example: Attach the TAM to a queue
+
+sai_attr.id = SAI_QUEUE_ATTR_TAM;
+sai_attr.value.oid = sai_tam_obj;
+
+sai_set_queue_attribute_fn(
+    sai_queue,
```

Comment:
Typo? Instead of sai\_queue, it should be sai\_queue\_obj used as SAI\_TAM\_COUNTER\_SUBSCRIPTION\_ATTR\_OBJECT\_ID earlier. 

If yes, this is a redundancy in this new API model.

Reply:
Yes, it is a typo (more like a missed edit when adding _obj to follow the style of other examples). 

Agree it is a redundancy. I also considered the alternative of just using the TAM attachment point at the switch as an on/off for a TAM defined for this kind of counter collection. What are your thoughts on that? 

In particular, it might increase extensibility to other objects which do not currently have a defined TAM bind-point (and be a bit easier to use since the telemetry can be enabled/disabled all-at-once if desired (rather than object-by-object)

Reply:
@JaiOCP , I would like to get your thoughts on this as well. Whether we can/should just use an attachment to switch for this kind of telemetry.